libparsec/crates/client/src/invite/greeter.rs

```diff
@@ -253,6 +263,129 @@ impl From<ConnectionError> for GreetInProgressError {
     }
 }
 
+// Cancel greeting attempt helper
+
+async fn cancel_greeting_attempt(
+    cmds: &AuthenticatedCmds,
+    greeting_attempt: GreetingAttemptID,
+    reason: CancelledGreetingAttemptReason,
+) -> Result<(), GreetInProgressError> {
+    use authenticated_cmds::latest::invite_greeter_cancel_greeting_attempt::{Rep, Req};
+
+    let req = Req {
+        greeting_attempt,
+        reason,
+    };
+    let rep = cmds.send(req).await?;
+
+    match rep {
+        Rep::Ok => Ok(()),
+        // Expected errors
+        Rep::InvitationCompleted => Err(GreetInProgressError::AlreadyDeleted),
+        Rep::InvitationCancelled => Err(GreetInProgressError::AlreadyDeleted),
+        Rep::AuthorNotAllowed => Err(GreetInProgressError::GreeterNotAllowed),
+        Rep::GreetingAttemptAlreadyCancelled {
+            origin,
+            reason,
+            timestamp,
+        } => Err(GreetInProgressError::GreetingAttemptCancelled {
+            origin,
+            reason,
+            timestamp,
+        }),
+        // Unexpected errors
+        Rep::GreetingAttemptNotFound => Err(anyhow::anyhow!("Greeting attempt not found").into()),
+        Rep::GreetingAttemptNotJoined => Err(anyhow::anyhow!("Greeting attempt not joined").into()),
+        rep @ Rep::UnknownStatus { .. } => {
+            Err(anyhow::anyhow!("Unexpected server response: {:?}", rep).into())
+        }
+    }
+}
+
+// Greeter step helper
+
+async fn run_greeter_step_until_ready(
+    cmds: &AuthenticatedCmds,
+    greeting_attempt: GreetingAttemptID,
+    greeter_step: invite_greeter_step::GreeterStep,
+    time_provider: &TimeProvider,
+) -> Result<invite_greeter_step::ClaimerStep, GreetInProgressError> {
+    let mut throttle = Throttle::new(time_provider);
+    let req = invite_greeter_step::Req {
+        greeting_attempt,
+        greeter_step,
+    };
+
+    // Loop over the requests
+    loop {
+        // Throttle the requests
+        throttle.throttle().await;
+
+        // Send the request
+        let rep = cmds.send(req.clone()).await?;
+
+        // Handle the response
+        return match rep {
```

Comment:
FYI you can break a loop with a return value

```rust
loop {
    break 2;
}

Reply:
Good to know, but I somehow find to existing more readable